flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -312,6 +323,16 @@ BufferAndBacklog pollBuffer() {
                     decreaseBuffersInBacklogUnsafe(bufferConsumer.isBuffer());
                 }
 
+                // if we have an empty finished buffer and the exclusive credit is 0, we just return
+                // the empty buffer so that the downstream task can release the allocated credit for
+                // this empty buffer, this happens in two main scenarios currently:
+                // 1. all data of a buffer builder has been read and after that the buffer builder
+                // is finished
+                // 2. in approximate recovery mode, a partial record takes a whole buffer builder
+                if (buffersPerChannel == 0 && bufferConsumer.isFinished()) {
+                    break;
+                }
+
```

Comment:
I can not seem to respond in the previous thread, so I need to start a new one.

> Let's maybe focus on the 3rd case first and we assume that the exclusive credit is 0.
>
> 1. There are only one data buffer in the queue.
> 2. Flush triggered.
> 3. All data of the first buffer is committed but the buffer is still not finished.
> 4. All data of the buffer is consumed by pollBuffer and the available credit becomes 0.
> 5. The first buffer is finished, the second event is added and the data available notification is triggered.
> 6. The upstream announces backlog to the downstream to request a credit.
> 7. The upstream receives available credit and start to pollBuffer.
> 8. Skip the first empty buffer and send the second event.
> 9. The downstream receive the event but the event does not consume any credit.
>
> Do you mean we should change the current logic and release the floating buffer for event in some cases (including reduce the available credit by 1 at the upstream, currently the available credit is not decreased for event)?

No, but I think we could send this regardless if any is credit available or not as we are doing right now. I think we are also already attaching information about the backlog to such event. One thing to add (unless we are not doing it already) would be to use this backlog information, to maybe release floating buffers if backlog dropped to 0?

> If there are multiple empty buffers, should we just skip the first one or should we skip all?

We could skip all of them, until we reach one of the three options:
1. non empty data buffer
2. event (check above)
3. last empty buffer, without any events after it - here we would indeed need to send that empty buffer

Reply:
> One thing to add (unless we are not doing it already) would be to use this backlog information, to maybe release floating buffers if backlog dropped to 0?

Currently, we are not doing that. Actually, I think it is a little complicated to do so. Because we need to keep consistency between the sender side available credit and the receiver side floating buffers. If we just release the floating buffers at the receiver side, if the sender side available credit is not reset, then there is may data sent out without buffers at receiver side to receive them. If we also reset the available credit at the sender side when the backlog is 0, there is a possibility that some AddCredit messages are on the way and we are not resetting this part. Maybe one way is to not sending any data out after sending a buffer with 0 backlog at sender side, then the receivers clear all floating credits and send a reset message to the senders. Then the senders reset all available credits. This process is similar to the channel blocking and resumption. I think this is a little complicated and can incur extra overhead.

What do you think? Or is there any simple way?

Reply:
Yes, you are right. It wouldn't be that simple. In that case, how complicated would it be to optimise the code to skip the all of the empty buffers until:
1. non empty data buffer
2. event (then send empty buffer first)
3. last empty buffer, without any events after it - here we would indeed need to send that empty buffer

?

Reply:
Currently, I can only come up with the following way, which depends on the downstream to reset the available credit of the upstream. This at least needs to add a special network message and propagating this message can incur extra overhead. If you think this is really important, I will spend some time to rethink about it and see if I can find a better way to solve it.

> Maybe one way is to not sending any data out after sending a buffer with 0 backlog at sender side, then the receivers clear all floating credits and send a reset message to the senders. Then the senders reset all available credits. This process is similar to the channel blocking and resumption.

Reply:
Ok, let's go with the current way for now, as with `buffersPerChannel == 0` check it's not that critical. But it would be nice if you could spend a little time thinking if we can improve this.

Reply:
OK, thanks a lot. I will spend some time thinking if we can improve it.